feat: checkpoint migrations

bdvllrs opened the pull request:
New features or refactoring of models might break model checkpoints.
As discussed during a @ecmwf/anemoitechnicalsubgroup meeting, this PR introduces a systematic way of migrating model checkpoints so that they can be always up-to-date with the anemoi version. This will avoid the hard choice of updating anemoi and retrain a checkpoint or keep an old anemoi version.

## The idea
The idea is to defer the responsability of keeping a working checkpoint to the person who introduces a breaking change.
This person only has to handle **his own** breaking change. This means that he can expect that checkpoint are working before.

This takes inspiration from database migration systems that apply changes to a database incrementally.

Specifically:
-  Checkpoints will have a new "migrations" key that store migrations that were already executed (if the key does not exist, it is assumed to not have any migration executed);
- when saving a checkpoint after training, the "migrations" key is populated with all existing migrations (no need to execute them);
-  if a PR introduces a change that breaks checkpoint, a new migration must be created with:
```bash
anemoi-models migration create my-migration-name
```
Here is an example migration file:
```python
from anemoi.models.migrations import CkptType
from anemoi.models.migrations import MigrationMetadata

metadata = MigrationMetadata(
    versions={
        "migration": "1.0.0",  # this tracks the version of the migration script. Might be useful if we want to refactor the API
        "anemoi-models": "0.8.1",  # this tracks the version of anemoi-models when this migration was created
    }
)


def migrate(ckpt: CkptType) -> CkptType:
    """Migrate the checkpoint"""
    ckpt["test"] = ckpt.pop("baz")
    return ckpt


def rollback(ckpt: CkptType) -> CkptType:
    """Rollback the migration"""
    ckpt["baz"] = ckpt.pop("test")
    return ckpt
```
This file contains some version metadata, a `migrate` function that should fix the introduced breaking change of the checkpoint, and a `rollback` function that does the opposite opperation.

## User interface
The user can use the command
```bash
anemoi-models migration sync ckpt_path [--steps STEPS]
```
to execute migrations of an old checkpoint. Only missing migrations will be executed.

Additional option:
- `--steps STEPS`: will execute only STEPS migration files. If negative, will rollback STEPS times

## The pipeline
- A new PR indroduces a change that breaks checkpoints
- A CI pipeline detects that checkpoints are broken by this PR
- A new migration file must be created and filled
- CI passes again
- The PR can be merged
- Users will migrate their checkpoint with `anemoi-models migrate`
 
## Roadmap
Here is the progress status of this PR (we may want to split this for the CI check?):
- [x] a script to generate new migrations
- [x] a system to select the migrations or rollbacks to execute
- [x] callback to store existing migrations when saving checkpoints
- [x] CI to detect that a PR broke a checkpoint and ask the developper to create a new migration. This can be a complement to #381.
- [x] add docs (see tests for now for examples)


## Open Questions
- Should the migrator be automatically ran if a checkpoint is out of date? Or should it always be done manually by the users? I tend to think that it should be manual, with anemoi logging the command to run to update the checkpoint when crashing.
- How to handle inference checkpoints? (probably regenerate the inference checkpoint from the complete model after migration)
- How to handle different model types? Should there be a set of migrations per models? Or a unique set of migrations but migrations also take some additional parameters to know the model to migrate? (although it can probably be infered from the checkpoint)

***As a contributor to the Anemoi framework, please ensure that your changes include unit tests, updates to any affected dependencies and documentation, and have been tested in a parallel setting  (i.e., with multiple GPUs). As a reviewer, you are also responsible for verifying these aspects and requesting changes if they are not adequately addressed. For guidelines about those please refer to https://anemoi.readthedocs.io/en/latest/***


<!-- readthedocs-preview anemoi-training start -->
----
📚 Documentation preview 📚: https://anemoi-training--386.org.readthedocs.build/en/386/

<!-- readthedocs-preview anemoi-training end -->

<!-- readthedocs-preview anemoi-graphs start -->
----
📚 Documentation preview 📚: https://anemoi-graphs--386.org.readthedocs.build/en/386/

<!-- readthedocs-preview anemoi-graphs end -->

<!-- readthedocs-preview anemoi-models start -->
----
📚 Documentation preview 📚: https://anemoi-models--386.org.readthedocs.build/en/386/

<!-- readthedocs-preview anemoi-models end -->

floriankrb commented on models/src/anemoi/models/commands/migrate.py:
```diff
@@ -0,0 +1,101 @@
+            ),
+        )
+        command_parser.add_argument(
+            "--n_migrations",
```
`n-migrations` instead of `n_migrations` please

bdvllrs replied:
done in 665f3ef

bdvllrs commented:
Here are some additional thoughts on handling multiple architectures:
I updated `versions` to `metadata` in migration files, and we could record some information such as on what type of model this migration is useful for. E.g.:
```python
metadata = Metadata(
    versions={
        "migration": "1.0.0",  # this tracks the version of the migration script. Might be useful if we want to refactor the API
        "anemoi-models": "0.8.1",  # this tracks the version of anemoi-models when this migration was created
    },
    architecture="graphtransformer",
)
```
We can add a filtering system on the migrator, to only apply migrations that follows certain constrains:
```python
migrator = Migrator(only={"architecture": ["graphtransformer"]})
```

gmertes reviewed (commented):
This is really cool stuff!

I do have one suggestion about the CLI: I suggest to consolidate CLI commands as much as possible. So instead of creating two distinct commands, put everything under 1 shared command and use subparsers for the different actions. Extra actions can then easily be added to the top level command.

I suggest something like:
```
anemoi-models migration create  (instead of create-migration)
anemoi-models migration apply   (instead of migrate)
```

Alternatively, `migrate` instead of `migration`.

We already use subparsers in many places, for example [here](https://github.com/ecmwf/anemoi-core/blob/fecc928470b3fd3662253b6cb575ad8c8b78c14a/training/src/anemoi/training/commands/mlflow.py#L24).

bdvllrs commented:
> One thought on the proposed pipeline. Some changes might break checkpoints to a degree where it does not make sense to patch them (an example would be some innovation that completely replaces parts of one of the model architectures), so requiring a migration file for PRs might be too strict? In such cases it would still be useful to somehow flag the change as breaking in the Migrator, and for the PR author to explain why & how.

Hi @havardhhaugen , this is a great idea! I added this feature with the "final" metadata in the migration. 
A final migration indicates that there is a change that cannot be migrated. It can be generated with:
```
anemoi-models migration create --final my-migration-name
```
and it looks like 
```python
from anemoi.models.migrations import CkptType
from anemoi.models.migrations import FinalMigrationException
from anemoi.models.migrations import MigrationMetadata

metadata = MigrationMetadata(
    versions={
        "migration": "1.0.0",
        "anemoi-models": "1.0.0",
    },
    final=True,
)


def migrate(ckpt: CkptType) -> CkptType:
    """Migrate the checkpoint."""
    raise FinalMigrationException


def rollback(ckpt: CkptType) -> CkptType:
    """Rollback the migration."""
    raise FinalMigrationException
```

The migrator will try to migrate up until the next final migration. If an explicit number of steps if provided, it will error out with a `FinalMigrationException`.

Similarly, too recent checkpoint cannot be rollbacked before a final migration.

bdvllrs commented:
@gmertes  Thanks for the comment!

> anemoi-models migration create  (instead of create-migration)
> anemoi-models migration apply   (instead of migrate)

I went for `anemoi-models migration create` and `anemoi-models migration sync` to follow the python API.


Additionnaly I removed all of the `--target` and `--n-migrations` options as I found they do not bring anything. It's easier to just pip install the desired version and call `anemoi-models migration sync`.

anaprietonem commented on models/docs/cli/migration.rst:
```diff
@@ -0,0 +1,66 @@
+
+   anemoi-models migration inspect PATH_TO_CKPT
+
+You can remove colors with the ``--no-color`` argument.
```
could be worth adding an example of that command to the docs?

bdvllrs replied:
done in 9c645715

anaprietonem commented on models/docs/contributing.rst:
```diff
@@ -18,3 +18,14 @@ anemoi-models include unit tests that can be executed locally using
+ Provide checkpoint Migrations
+*******************************
+
+If your changes break existing checkpoints, you must provide a
```
Could be worth mentioning here how one can detect if the changes break the checkpoint - ie mention the tests that  checks that?

mchantry replied:
I don't think it's documented but Vera can confirm. Adding a quick mention to this test would be a nice addition.

anaprietonem commented on models/docs/cli/migration.rst:
```diff
@@ -0,0 +1,66 @@
+
+.. code:: bash
+
+   anemoi-models migration sync PATH_TO_CKPT
```
Not related strictly to the PR but the CLI in training also have some utilities related to checkpoints - https://github.com/ecmwf/anemoi-core/blob/main/training/src/anemoi/training/commands/checkpoint.py to convert a training checkpoint to an inference one. It could be good in the future to think about unifying these two CLIs (note the training conversion requires pytorch lightning so that's why it lives in anemoi-training and not in models)

anaprietonem commented on models/docs/create-migrations.rst:
```diff
@@ -0,0 +1,319 @@
+ Full example
+**************
+
+Here is a full example of a migration to fix `PR 433
```
Thanks for this example! very nice

anaprietonem commented on models/docs/create-migrations.rst:
```diff
@@ -0,0 +1,319 @@
+          A MigrationContext instance
+       """
+
+To generate your script with the setup callbacks, use the
```
it should be possible to create a single migration script that handles both. the case of a layer renaming and this one right?
I think it could be worth adding a bit more here in regards to when are setup callbacks required and when 'renaming' are just enough

bdvllrs replied:
Yes absolutely! I'll updates the docs to make it clear.

anaprietonem commented on models/src/anemoi/models/migrations/1750845283_test.py:
```diff
@@ -0,0 +1,29 @@
+# (C) Copyright 2024 Anemoi contributors.
```
where is the numbers for these files comign from?

bdvllrs replied:
It's a timestamp of the time when `anemoi-models migration create` is triggered. This is used to keep them ordered.

There has been a concern brought up during an anemoi community meeting that this could be problematic when merging different PRs with migrations (the final order should depend on merging order, not on time of script creation).

Honestly, this is still an open question: should there be a CI test that checks that ordering is correct before merging?

anaprietonem commented on models/src/anemoi/models/migrations/__init__.py:
```diff
@@ -0,0 +1,691 @@
+        # Compatibility groups. Checkpoints cannot be migrated past their
+        # own group. This is useful to indicate when migrating checkpoints is no longer
+        # supported.
+        self._grouped_migrations: list[list[Migration]] = []
```
How could someone pass a migrations that are not in the current folder?

anaprietonem replied:
No I was just curious, i think it's better if that's removed cause I assume for debugging/understanding the flow knowing that the sync/create commands are looking at just one folder is clearer. Thanks!

anaprietonem commented on models/src/anemoi/models/migrations/__init__.py:
```diff
@@ -0,0 +1,691 @@
+                self._grouped_migrations.append(current_group)
+                current_group = []
+            current_group.append(migration)
+        self._grouped_migrations.append(current_group)
```
could we add a logger to print here a 'summary' of the migrations found?

anaprietonem replied:
what's the other in which these migrations are applied?

bdvllrs commented on models/src/anemoi/models/migrations/__init__.py:
```diff
@@ -0,0 +1,691 @@
+LOGGER = logging.getLogger(__name__)
+
+
+class MissingMigrationException(BaseException):
```
This is triggered when you try to execute a migration, but the implementation is missing in one of the scripts.
This allows having migration only scripts (no rollback) (and techically rollback only and no migrate, but I don't see how this is useful).

bdvllrs commented on models/docs/create-migrations.rst:
```diff
@@ -0,0 +1,319 @@
+
+   metadata = MigrationMetadata(
+       versions={
+           "migration": "1.0.0",
```
I added a not in the docs that the metadata object is automatically generated and that contributors should not update it.

bdvllrs commented on models/docs/create-migrations.rst:
```diff
@@ -0,0 +1,319 @@
+
+.. code:: bash
+
+   anemoi-models migration create --final MIGRATION_NAME
```
Not exactly. This is not for end users, but for a contributor that introduced a feature that cannot be migrated (e.g. a new layer is added, and a full retrain is needed to compute the missing weights). Then he has to generate a "final" migration that acts as a separator (checkpoint trained before this migration cannot be migrated passed it).

Until now, the API you describe doesn't exist. I added in 3f447d4 the fisrt incompatible anemoi-models version information in the error message when you try to sync a checkpoint too old. Also the inspect command now also shows it.

bdvllrs commented on training/src/anemoi/training/diagnostics/callbacks/__init__.py:
```diff
@@ -195,6 +196,7 @@ def get_callbacks(config: DictConfig) -> list[Callback]:
         (
```
I'm not sure what value the comments bring compared to the following code, but ok.

bdvllrs commented on models/pyproject.toml:
```diff
@@ -43,8 +43,10 @@ dynamic = [ "version" ]
 dependencies = [
   # Fix certain dependencies during development
   "anemoi-utils>=0.1.9",
+  "cloudpickle>=3.1",
```
done in dcd1f35e

bdvllrs commented on models/docs/create-migrations.rst:
```diff
@@ -0,0 +1,319 @@
+           "anemoi.models.schemas.data_processor.DataSchema", "anemoi.training.schemas.data.NormalizerSchema"
+       )
+
+
```
done in 2a504c16

bdvllrs commented on models/src/anemoi/models/migrations/__init__.py:
```diff
@@ -0,0 +1,691 @@
+            current_group.append(migration)
+        self._grouped_migrations.append(current_group)
+
+        self._raise_missing_migrations = raise_missing_migrations
```
good catch, this is old code. Removing.

bdvllrs commented on models/src/anemoi/models/migrations/__init__.py:
```diff
@@ -0,0 +1,691 @@
+        if migrations is None:
+            migrations = _migrations_from_path(MIGRATION_PATH, __name__)
+
+        # Compatibility groups. Checkpoints cannot be migrated past their
```
This is more of an implementation detail. It just groups together migrations that are compatible with one another (in between "final" migrations). This is to detect if the checkpoint is compatible with the current version, or if one need to use an older version of anemoi-models.

Is this explanation clearer? Should I rephrase with this the comment in the code?

anaprietonem replied:
Okey that's clear now but I think updating the comment in the code could be good - a follow up question is then let' assume we have migration1, migration2, migration3(final), migration4, migration5(filnal) and migration 6
where 6 is newer and migration1 is the oldest. Then in which order is this applied? It's related to the question above about the order.

bdvllrs commented on models/src/anemoi/models/commands/migration.py:
```diff
@@ -0,0 +1,372 @@
+        imports_items.append("from anemoi.models.migrations import MigrationMetadata")
+        imports = "\n".join(imports_items)
+
+        with open(args.path / name, "w") as f:
```
Yes, it's probably better. I chose against it at the beginning because the generation script was easy, but it's getting too messy.

bdvllrs commented on models/src/anemoi/models/commands/migration.py:
```diff
@@ -0,0 +1,372 @@
+    return text
+
+
+class Migration(Command):
```
Added the error in 2a504c1.

bdvllrs commented:
Thanks @anaprietonem  for your extensive review! I've answered back to your comments.

anaprietonem commented on models/tests/.gitignore:
```diff
@@ -0,0 +1 @@
+!checkpoints/
```
if you don't add this, this should be tracked by default?

anaprietonem commented on models/src/anemoi/models/commands/migration.py:
```diff
@@ -0,0 +1,431 @@
+            new_name = f"{new_timestamp + k}_{new_name}"
+            print(f"Renaming {name} to {new_name}.")
+            print(path, path.with_name(new_name))
+            # path.rename(path.with_name(new_name))
```
I guess this can be removed?

anaprietonem commented on models/src/anemoi/models/commands/migration.py:
```diff
@@ -0,0 +1,431 @@
+            path = MIGRATION_PATH / name
+            _, _, new_name = name.partition("_")
+            new_name = f"{new_timestamp + k}_{new_name}"
+            print(f"Renaming {name} to {new_name}.")
```
I noticed there is some mixing of print and LOGGER statements. I'd propose to replace the print statements and just use the LOGGER?

anaprietonem commented:
> @anaprietonem I've updated the docs, hopefully things are more clear now. Here are some other changes that I've made to help you review:
> 
> * I've moved the scripts in `anemoi.models.migrations.scripts` so that they are in a separate folder
> * I've added a history of operations done in the checkpoint in the metadata under the "migrations.history" key. This will allow better debugging as it traces exactly what happenned to the checkpoint
> * I've added a "signature" key stored in the migration. This is a sha256 digest of the migration script (passed through the ast package so that it's indentation and comment independant). It will also raise a warning to the user if the signature of scripts registered in the checkpoint are different from the signature of the current script. I did this to allow us updating old scripts if we find there is a bug. People that already executed these script will then get a warning asking them to reapply the migrations.
> * the `rollback_setup` callback can automatically be inferred from the `migrate_setup` callback, so I did that.
> 
> What's missing now:
> 
> * the CI part to update the anemoi-models version in scripts.
> * update the `migration create` command to put a placeholder for the version.

Very nice Benjamin! thanks for addressing the comments and for the summary. I think this page looks very nice https://anemoi--386.org.readthedocs.build/projects/models/en/386/modules/migrations.html . Only thing is the two sections at the bottom are those expected to be autocompleted?

from the what's missing - the CI parts - that I guess refer to the branch from Gert and that's already merged - so are those two points still applicable?

bdvllrs commented:
> Very nice Benjamin! thanks for addressing the comments and for the summary. I think this page looks very nice https://anemoi--386.org.readthedocs.build/projects/models/en/386/modules/migrations.html . Only thing is the two sections at the bottom are those expected to be autocompleted?

Yes, it's supposed to be the API from automodule, but it seems like it's not generated, I'll look into why. (Maybe it needs a __all__ export?)

> from the what's missing - the CI parts - that I guess refer to the branch from Gert and that's already merged - so are those two points still applicable?

Yes, those are resolve already. From my side, everything is great!

The only thing I've realized: I'm checking the signature on the registered migrations, which means that checkpoint who were trained after the migration and did not ran it might get a warning if migrations change. I should probably do it from the metadata history field which actually tracks the migration scripts that were executed.

anaprietonem reviewed (approved):
Great work @bdvllrs! thanks for addressing the feedback, the PR LGTM!
